### ISSUE TITLE
Proposal: Allow storing config in os.Env (environment Variables)

### ISSUE THREAD

**Issue by Longwater1234** [Golang-Tanzania/GoPesa#10]
Instead of using config.json as it is currently, it would be very useful to be able to store the config values (PRIVATE and PUBLIC keys) from `os.GetEnv` see [official docs](https://pkg.go.dev/os#Getenv), which provides many ways to be able to securely GET config-values as follows: 
1.  For example, if used in web application that already uses an existing `.env` file, the user of this Library can simply append PRIVATE_KEY and PUBLIC_KEY to the `.env` file. (make sure it  is gitignored)
2. If running from a docker container, a user of this library can simply append `--env PRIVATE_KEY=xxxxx, ...` to the `docker run` command
3. If run from an IDE , a user can simply store the config in the IDE environmental variables settings panel (all major IDE's support this).
4. If run from the Terminal (when quickly testing), a user can simply pass the config values as Unix shell comands before starting the application. 

5. When run an application from a Cloud Provider, a user can set the env in the Cloud dashboard of host they are using, together with other app secrets. 

As you can see, a user can either store the config in a `.env` file or simply pass as command args to the application.

**Comment by Longwater1234** [comment]
Or, a simpler alternative, just modify the public `gopesa.initialize()` function to pass parameters: `privateKey` and `publicKey`. Then the developer (user of this library) will decide on their own how to store & load the keys

**Comment by Hopertz** [comment]
@Longwater1234  The first solution seems to be the most common one to me.

**Comment by Longwater1234** [comment]
@Hopertz Alright, i suggest we let the developer decide how he/she wants to store the `env` keys (option 2). Whether as json, args or .env file, etc..

So therefore, we just modify the `initialize()` function to recieve both keys as String.

**Comment by Hopertz** [comment]
@Longwater1234 so we won't enforce that on the developer; that's fine.

**Comment by Hopertz** [comment]
> Instead of using config.json as it is currently, it would be very useful to be able to store the config values (PRIVATE and PUBLIC keys) from `os.GetEnv` see [official docs](https://pkg.go.dev/os#Getenv), which provides many ways to be able to securely GET config-values as follows:
> 
>     1. For example, if used in web application that already uses an existing `.env` file, the user of this Library can simply append PRIVATE_KEY and PUBLIC_KEY to the `.env` file. (make sure it  is gitignored)
> 
>     2. If running from a docker container, a user of this library can simply append `--env PRIVATE_KEY=xxxxx, ...` to the `docker run` command
> 
>     3. If run from an IDE , a user can simply store the config in the IDE environmental variables settings panel (all major IDE's support this).
> 
>     4. If run from the Terminal (when quickly testing), a user can simply pass the config values as Unix shell comands before starting the application.
> 
>     5. When run an application from a Cloud Provider, a user can set the env in the Cloud dashboard of host they are using, together with other app secrets.
> 
> 
> As you can see, a user can either store the config in a `.env` file or simply pass as command args to the application.

@Longwater1234 Kindly check the latest implementation and give your opinions.

**Comment by Longwater1234** [comment]
@Hopertz Alright, I'll take a look today

**Comment by Longwater1234** [comment]
@Hopertz So I have taken a close look at the rewritten library. It's very well organized; great job 👍! I have created another issue #16  concerning structs;, you can take a look. But concerning this current issue of ENV, I think the current solution is still not optimal or user-friendly. Here's why: 
1. It adds a third-party library like Viper just to load the env variables. Adds extra weight to this library.
2. The init function is still not flexible depending on the deployment environment (in your `mpesa.Client.LoadKeys`, I must pass **4 DIFFERENT variables** just to load keys). It's more complex than before! And if I change my deployment environment, I have to modify my application source code again. Sadly, very tight coupling of App source code and ENV. Also it only considers `file` config alone, ignoring reading from OS.ENV or args.

It would be easier to maybe have a single init function `mpesa.NewClient( ...)` or your same `client.LoadKeys( )`  that accepts directly `public_key` and `api_key`, AND mpesa.Env as `string`. We should let **the Developer decide** how they are going to _store & load_ the strings, whether using their own 3rd party libraries, from file or args, however they want. **We can still OFFER SUGGESTIONS** in README.md on how to **safely** store the config, but let's not be **TOO STRICT** on how they should be loaded. This library may be used in many different environments (large and small) and devices, and each developer or environment has their own way of loading the env Keys. Let's not be too restrictive on general preferences.

This is just my opinion. I hope you consider them. Thanks.

**Comment by Hopertz** [comment]
> @Hopertz So I have taken a close look at the rewritten library. It's very well organized; great job 👍! I have created another issue #16 concerning structs;, you can take a look. But concerning this current issue of ENV, I think the current solution is still not optimal or user-friendly. Here's why: 
> 
>     1. It adds a third-party library like Viper just to load the env variables. Adds extra weight to this library.
> 
>     2. The init function is still not flexible depending on the deployment environment (in your `mpesa.Client.LoadKeys`, I must pass **4 DIFFERENT variables** just to load keys). It's more complex than before! And if I change my deployment environment, I have to modify my application source code again. Sadly, very tight coupling of App source code and ENV. Also it only considers `file` config alone, ignoring reading from OS.ENV or args.
> 
> 
> It would be easier to maybe have a single init function `mpesa.NewClient( ...)` or your same `client.LoadKeys( )` that accepts directly `public_key` and `api_key`, AND mpesa.Env as `string`. We should let **the Developer decide** how they are going to _store & load_ the strings, whether using their own 3rd party libraries, from file or args, however they want. **We can still OFFER SUGGESTIONS** in README.md on how to **safely** store the config, but let's not be **TOO STRICT** on how they should be loaded. This library may be used in many different environments (large and small) and devices, and each developer or environment has their own way of loading the env Keys. Let's not be too restrictive on general preferences.
> 
> This is just my opinion. I hope you consider them. Thanks.


   Thank you for your time.

1. Using 3rd party library viper has enabled loading keys from YAML, JSON and .ENV files and also that you just need to pass path to dir contain the file ,  file name and extension.The name config does have to appear you can name anything you want. This gives the developer flexibility of using either of the file extension to store their keys and store  in any path. The Key thing i missed is that i did not add put option of using args.

2.  I agree viper adds extra weight to the library. We can implement our own shouldn't  be hard given that we just need few functionalities.

3. Did not get you on the last comment can you explain more?

**Comment by Longwater1234** [comment]
@Hopertz I mean no need to re-implement our own either. All i mean the current solution is too strict on using _File-based_ config. For example, Some developers may want to load their Mpesa config along with other ENV vars of their web app (database passw, url etc). They may not store it in FILE, they may use Operating system Variables or Docker env, or ARGS ENV variables. NOT FILE BASED (`.env, yaml, json`). 

Additionally, All major and minor cloud hosting providers also dont offer file based loading of config Env values, they offer OS-based/container-based ENV, on the cloud Admin console, where you can fill in all your Application Env variables. (That is, they expect you to `os.GetEnv` in Golang). And the values are encrypted automatically.

So in short, the init function should just recieve 3 params: `publicKey`, `apiKey` as string and `mpesa.Envtype` (ie Sandbox or Prod). This will allow all 3 options: file-based, args, or OS /container based variables

**Comment by Hopertz** [comment]
> @Hopertz I mean no need to re-implement our own either. All i mean the current solution is too strict on using _File-based_ config. For example, Some developers may want to load their Mpesa config along with other ENV vars of their web app (database passw, url etc). They may not store it in FILE, they may use Operating system Variables or Docker env, or ARGS ENV variables. NOT FILE BASED (`.env, yaml, json`).
> 
> Additionally, All major and minor cloud hosting providers also dont offer file based loading of config Env values, they offer OS-based/container-based ENV, on the cloud Admin console, where you can fill in all your Application Env variables. (That is, they expect you to `os.GetEnv` in Golang). And the values are encrypted automatically.
> 
> So in short, the init function should just recieve 3 params: `publicKey`, `apiKey` as string and `mpesa.Envtype` (ie Sandbox or Prod). This will allow all 3 options: file-based, args, or OS /container based variables

Great, Waiting for your pull request on this.

**Comment by Hopertz** [comment]
@Longwater1234   also lets have mpesa.NewClient(....) remove mpesa.LoadKeys(..)

**Comment by Longwater1234** [comment]
@Hopertz  Alright i will work on it this weekend. By this Sunday it will be ready.

**Comment by Hopertz** [comment]
@Longwater1234 decided to fix asap, have look at it.

**Comment by Longwater1234** [comment]
@Hopertz Seen. 3 final touches
1. You forgot PUBLIC_KEY in the contructor.
2. Since you are now using `context.Background()` (which is GREAT), instead of user creating & passing their own httpClient with timeout, they can directly pass their own `context.WithTimeout{}` on the fly, (which has more capabilities). **So NO need of having httpClient setup externally** Optionally, If they dont want to pass `ctx.WithTimeout`, they can just pass `context.Background()` as default. Let's not force all users use ctx.Background.
3. For session timeout in `mpesa.NewClient("your-api-key", "sandbox", 24)`, instead of `int` type, let's pass type `time.Duration`. So the result will be mpesa.NewClient("your-api-key", "sandbox", 24 * time.Hour). Or can restrict to Hour only. This is direct type-safety and validilty check.

**Comment by Longwater1234** [comment]
@Hopertz Also, the env type, "sandbox" or "production" can be enum

**Comment by Hopertz** [comment]
@Longwater1234  I am disappointed, i didn't  forget the  public key, how could the code run?
Also context has other use case not only  ctx.WithTimeout (). Both have importance in the code. Using enums for two constants is an overkill.

**Comment by Longwater1234** [comment]
@Hopertz yes, i meant with context, one can achieve 2 main goals with 1 shot, both **timeout** and all other context-specific uses. So this may remove the need of manually setting up httpClient.

**Comment by Hopertz** [comment]
@Longwater1234 Let us give the developer the freedom. He/she can decide what to use and not limit on how to do things.
